File: packages/sdk-python/devbolt/engine.py
"""Main engine for feature flag evaluation."""

import time
from typing import Optional

from .evaluator import FlagEvaluator
from .logger import NoOpLogger
from .parser import ConfigParser
from .types import (
    EvaluationContext,
    EvaluationResult,
    EvaluationMetadata,
    FlagConfig,
    FlagNotFoundError,
    FlagsConfig,
    Logger,
)


class FlagEngine:
    """Main engine for feature flag evaluation."""

    def __init__(
        self,
        config: FlagsConfig,
        logger: Optional[Logger] = None,
        strict: bool = False,
    ):
        """Initialize flag engine.

        Args:
            config: Flags configuration
            logger: Optional logger instance
            strict: Enable strict mode (throw on missing flags)
        """
        self._config = config
        self.logger = logger or NoOpLogger()
        self.strict = strict
        self.evaluator = FlagEvaluator(self.logger)

        self.logger.info(
            "FlagEngine initialized",
            {
                "flag_count": len(config),
                "strict": strict,
            },
        )

    @classmethod
    def from_yaml(
        cls,
        yaml_content: str,
        logger: Optional[Logger] = None,
        strict: bool = False,
    ) -> "FlagEngine":
        """Create engine from YAML string.

        Args:
            yaml_content: YAML configuration content
            logger: Optional logger instance
            strict: Enable strict mode

        Returns:
            FlagEngine instance
        """
        config = ConfigParser.parse_yaml(yaml_content)
        return cls(config, logger=logger, strict=strict)

    @classmethod
    def from_file(
        cls,
        file_path: str,
        logger: Optional[Logger] = None,
        strict: bool = False,
    ) -> "FlagEngine":
        """Create engine from file.

        Args:
            file_path: Path to config file
            logger: Optional logger instance
            strict: Enable strict mode

        Returns:
            FlagEngine instance
        """
        config = ConfigParser.parse_file(file_path)
        return cls(config, logger=logger, strict=strict)

    def evaluate(
        self,
        flag_name: str,
        context: Optional[EvaluationContext] = None,
    ) -> EvaluationResult:
        """Evaluate a feature flag.

        Args:
            flag_name: Name of the flag
            context: Optional evaluation context

        Returns:
            EvaluationResult

        Raises:
            FlagNotFoundError: If flag not found and strict mode enabled
        """
        flag_config = self._config.get(flag_name)

        if not flag_config:
            if self.strict:
                raise FlagNotFoundError(flag_name)

            self.logger.warn(f"Flag '{flag_name}' not found, returning disabled")
            return EvaluationResult(
                flag_name=flag_name,
                enabled=False,
                reason="Flag not found",
                metadata=EvaluationMetadata(timestamp=time.time()),
            )

        eval_context = context or EvaluationContext()
        return self.evaluator.evaluate(flag_name, flag_config, eval_context)

    def is_enabled(
        self,
        flag_name: str,
        context: Optional[EvaluationContext] = None,
    ) -> bool:
        """Check if a flag is enabled.

        Args:
            flag_name: Name of the flag
            context: Optional evaluation context

        Returns:
            True if enabled, False otherwise
        """
        return self.evaluate(flag_name, context).enabled

    def get_all_flags(self) -> list[str]:
        """Get all flag names.

        Returns:
            List of flag names
        """
        return list(self._config.keys())

    def get_flag_config(self, flag_name: str) -> Optional[FlagConfig]:
        """Get flag configuration.

        Args:
            flag_name: Name of the flag

        Returns:
            FlagConfig if found, None otherwise
        """
        return self._config.get(flag_name)

    def get_config(self) -> FlagsConfig:
        """Get entire configuration.

        Returns:
            Copy of flags configuration
        """
        return self._config.copy()

    def update_config(self, config: FlagsConfig) -> None:
        """Update configuration (for hot reloading).

        Args:
            config: New flags configuration
        """
        self._config = config
        self.logger.info(
            "Configuration updated",
            {"flag_count": len(config)},
        )

File: packages/core/src/engine.ts
import type {
  FlagsConfig,
  EvaluationContext,
  EvaluationResult,
  EngineOptions,
  Logger,
} from "./types.js";
import { FlagNotFoundError } from "./types.js";
import { ConfigParser } from "./parser.js";
import { FlagEvaluator } from "./evaluator.js";
import { NoOpLogger } from "./logger.js";

/**
 * Main engine for feature flag evaluation
 */
export class FlagEngine {
  private config: FlagsConfig;
  private readonly evaluator: FlagEvaluator;
  private readonly logger: Logger;
  private readonly strict: boolean;

  constructor(config: FlagsConfig, options: EngineOptions = {}) {
    this.config = config;
    this.logger = options.logger || new NoOpLogger();
    this.strict = options.strict ?? false;
    this.evaluator = new FlagEvaluator(this.logger);

    this.logger.info("FlagEngine initialized", {
      flagCount: Object.keys(config).length,
      strict: this.strict,
    });
  }

  /**
   * Create engine from YAML string
   */
  static fromYAML(yaml: string, options?: EngineOptions): FlagEngine {
    const config = ConfigParser.parseYAML(yaml);
    return new FlagEngine(config, options);
  }

  /**
   * Create engine from file
   */
  static fromFile(filePath: string, options?: EngineOptions): FlagEngine {
    const config = ConfigParser.parseFile(filePath);
    return new FlagEngine(config, options);
  }

  /**
   * Evaluate a feature flag
   */
  evaluate(
    flagName: string,
    context: EvaluationContext = {}
  ): EvaluationResult {
    const flagConfig = this.config[flagName];

    if (!flagConfig) {
      if (this.strict) {
        throw new FlagNotFoundError(flagName);
      }

      this.logger.warn(`Flag "${flagName}" not found, returning disabled`);
      return {
        flagName,
        enabled: false,
        reason: "Flag not found",
        metadata: { timestamp: Date.now() },
      };
    }

    return this.evaluator.evaluate(flagName, flagConfig, context);
  }

  /**
   * Check if a flag is enabled
   */
  isEnabled(flagName: string, context: EvaluationContext = {}): boolean {
    return this.evaluate(flagName, context).enabled;
  }

  /**
   * Get all flag names
   */
  getAllFlags(): string[] {
    return Object.keys(this.config);
  }

  /**
   * Get flag configuration
   */
  getFlagConfig(flagName: string) {
    return this.config[flagName];
  }

  /**
   * Get entire configuration
   */
  getConfig(): Readonly<FlagsConfig> {
    return Object.freeze({ ...this.config });
  }

  /**
   * Update configuration (for hot reloading)
   */
  updateConfig(config: FlagsConfig): void {
    this.config = config;
    this.logger.info("Configuration updated", {
      flagCount: Object.keys(config).length,
    });
  }
}
